warp10/src/main/java/io/warp10/continuum/gts/MetadataTextComparator.java

```diff
@@ -24,158 +24,213 @@
 import io.warp10.continuum.store.thrift.data.Metadata;
 
 /**
- * Sort Metadata according to text order (on name then on labels)
- * 
+ * Sort Metadata according to field values (labels, attributes or classname if null provided)
  */
 public class MetadataTextComparator implements Comparator<Metadata> {
 
   private final List<String> fields;
-  
+
   public MetadataTextComparator(List<String> fields) {
     this.fields = fields;
   }
-  
+
   @Override
   public int compare(Metadata o1, Metadata o2) {
-    
+
     if (null != this.fields && !this.fields.isEmpty()) {
       return compareWithFields(o1, o2);
     }
-        
+
     if (null == o1) {
       return -1;
     }
-    
+
     if (null == o2) {
       return 1;
     }
-    
+
     String name1 = o1.getName();
     String name2 = o2.getName();
-    
+
     if (null == name1) {
       return -1;
     }
-    
+
     if (null == name2) {
       return 1;
     }
-    
+
     int comp = name1.compareTo(name2);
-    
+
     if (0 != comp) {
       return comp;
     }
-   
+
+    //
+    // Names are identical, compare labels
+    //
+
     if (0 == o1.getLabelsSize() && 0 == o2.getLabelsSize()) {
       return 0;
     }
-    
+
     if (0 == o1.getLabelsSize()) {
       return -1;
     }
-    
+
     if (0 == o2.getLabelsSize()) {
       return 1;
     }
-    
-    //
-    // Names are identical, compare labels
-    //
-    
+
     List<String> labels1 = new ArrayList<String>(o1.getLabelsSize());
     labels1.addAll(o1.getLabels().keySet());
-    
+
     List<String> labels2 = new ArrayList<String>(o2.getLabelsSize());
     labels2.addAll(o2.getLabels().keySet());
-    
+
     Collections.sort(labels1);
     Collections.sort(labels2);
-    
-    int idx1 = 0;
-    int idx2 = 0;
-    
-    while(idx1 < labels1.size() && idx2 < labels2.size()) {
-      
+
+    int idx = 0;
+
+    while (idx < labels1.size() && idx < labels2.size()) {
+
       // Compare label names
-      comp = labels1.get(idx1).compareTo(labels2.get(idx2));
-      
+      comp = labels1.get(idx).compareTo(labels2.get(idx));
+
       if (0 != comp) {
         return comp;
       }
-      
+
       // Compare label values
-      comp = o1.getLabels().get(labels1.get(idx1)).compareTo(o2.getLabels().get(labels2.get(idx2)));
-      
+      comp = o1.getLabels().get(labels1.get(idx)).compareTo(o2.getLabels().get(labels2.get(idx)));
+
       if (0 != comp) {
         return comp;
       }
-      
-      //
+
       // Advance labels
-      //
-      
-      idx1++;
-      idx2++;
+      idx++;
     }
 
-    if (idx1 == idx2) {
+    //
```

Comment:
The original loop condition had an issue, but the original idea was that a GTS with fewer labels than another one would appear before. Your change does not adhere to this as the GTS are considered equal when they have the same labels for the smaller set of labels of the two.

Reply:
Done